### PR TITLE
feat/configuration-contains-value

### PR DESCRIPTION
**Pull request by rikkarth**
I have been working on a new small feature which aims to validate if a value is present in a configuration through a first match validation.

This is useful when looking for specific values in multiple configurations and often can be leveraged to return the right Configuration if a value is present.

It can also be used to validate the integrity of a Configuration.

It is natively present in java.util.Properties (see [JavaDoc](https://docs.oracle.com/javase/8/docs/api/java/util/Hashtable.html#containsValue-java.lang.Object-)).

I tried to mimic the structure already implemented for `containsKey` but added different implementation in order to retrieve the match for a value.

The first tests pass for all Configuration types, but I want to add more tests after the first appreciation of this PR.

I used recursion in some implementations for some Configuration types to retrieve the match for a value.

```java
    @Override
    protected boolean containsValueInternal(final String value) {
        return contains(getKeys(), value);
    }

    private boolean contains(Iterator<String> keys, final String value) {
        if (keys.hasNext()) {
            String nextKey = keys.next();
            Object valueFromKey = getProperty(nextKey);

            if (valueFromKey.equals(value)) {
                return true;
            }

            contains(keys, value);
        }

        return false;
    }
```

Tests currently fail here but only on GitHub, when I test them locally they don't fail. I would appreciate some feedback on this if possible.
image

```
Error:  Failures: 
Error:    TestServletConfiguration>TestAbstractConfiguration.testContainsValue:84 should return true for class class org.apache.commons.configuration2.web.TestServletConfiguration ==> expected: <true> but was: <false>
Error:    TestServletContextConfiguration>TestAbstractConfiguration.testContainsValue:84 should return true for class class org.apache.commons.configuration2.web.TestServletContextConfiguration ==> expected: <true> but was: <false>
Error:    TestServletFilterConfiguration>TestAbstractConfiguration.testContainsValue:84 should return true for class class org.apache.commons.configuration2.web.TestServletFilterConfiguration ==> expected: <true> but was: <false>
```

I'm happy for any feedback.

### CODE REVIEW

**Comment by garydgregory** [comment]
@rikkarth 
Make sure you run `mvn` by itself to verify that all checks pass.

**Comment by rikkarth** [comment]
This is the error when running `mvn`. Is this some build specification I've missed?

```
[ERROR] Failed to execute goal com.github.siom79.japicmp:japicmp-maven-plugin:0.20.0:cmp (default) on project commons-configuration2: There is at least one incompatibility: org.apache.commons.configuration2.AbstractConfiguration.containsValueInternal(java.lang.String):METHOD_ABSTRACT_ADDED_TO_CLASS,org.apache.commons.configuration2.ImmutableConfiguration.containsValue(java.lang.String):METHOD_ADDED_TO_INTERFACE -> [Help 1]
```

**Comment by rikkarth** [comment]
Ok, so I'm having an error that I think is build related:

Can you help correct this issue?

```
[ERROR] Failed to execute goal com.github.siom79.japicmp:japicmp-maven-plugin:0.20.0:cmp (default) 
on project commons-configuration2: There is at least one incompatibility: 
org.apache.commons.configuration2.AbstractConfiguration.containsValueInternal(java.lang.String):METHOD_ABSTRACT_ADDED_TO_CLASS,
org.apache.commons.configuration2.ImmutableConfiguration.containsValue(java.lang.String):METHOD_ADDED_TO_INTERFACE
```

**Comment by garydgregory** [comment]
Hello @rikkarth 
It looks like this PR breaks binary compatibility based on the japicmp error you quote and there are also failing unit tests.

Your profile states you are a "Problem-solving Prodigy", so... either less boasting or more problem solving ;-)

See also:
- https://garygregory.wordpress.com/2020/06/14/how-we-handle-binary-compatibility-at-apache-commons/ 
- https://wiki.openjdk.org/display/csr/Kinds+of+Compatibility
- https://docs.oracle.com/javase/specs/jls/se8/html/jls-13.html

HTH

**Comment by rikkarth** [comment]
@garydgregory , thank you for the feedback and providing the documentation.

I must rethink my profile description.

File: src/main/java/org/apache/commons/configuration2/DatabaseConfiguration.java
Comment (Claudenw) [inline src/main/java/org/apache/commons/configuration2/DatabaseConfiguration.java:487]: If I read the SQL_GET_PROPERTY correctly as well as the what you are passing in as arguments, you have used up all the prepared statement parameters and the `true` value in this line should be `false`.  

**WARNING** I may be mistaken.

File: src/main/java/org/apache/commons/configuration2/JNDIConfiguration.java
Comment (Claudenw) [inline src/main/java/org/apache/commons/configuration2/JNDIConfiguration.java:167]: See comments for similar method above.

File: src/main/java/org/apache/commons/configuration2/web/BaseWebConfiguration.java
Comment (Claudenw) [inline src/main/java/org/apache/commons/configuration2/web/BaseWebConfiguration.java:90]: See comments for similar method above.  

Do all of these classes implement a common interface or class where a common method might be placed?

**Review by Claudenw — CHANGES_REQUESTED** [review]
Thanks for your contribution.  There are a couple of issues I think need to be fixed.  You might also wait for another review before you make these changes.

File: src/main/java/org/apache/commons/configuration2/web/BaseWebConfiguration.java
Comment (rikkarth) [reply src/main/java/org/apache/commons/configuration2/web/BaseWebConfiguration.java]: They do, I reduced the duplication in my next commit.

File: src/main/java/org/apache/commons/configuration2/AbstractHierarchicalConfiguration.java
Comment (rikkarth) [reply src/main/java/org/apache/commons/configuration2/AbstractHierarchicalConfiguration.java]: 100% agree. Will perform the change.

Thank you.

File: src/main/java/org/apache/commons/configuration2/AbstractHierarchicalConfiguration.java
Comment (rikkarth) [reply src/main/java/org/apache/commons/configuration2/AbstractHierarchicalConfiguration.java]: That was my initial idea (use the while). The functional part of me decided against KISS.

I will revert to a while and no need for suplier since method (now updated in last commit) is only called from one place.

File: src/main/java/org/apache/commons/configuration2/DatabaseConfiguration.java
Comment (rikkarth) [reply src/main/java/org/apache/commons/configuration2/DatabaseConfiguration.java:487]: Turned it off but defined a test to check for return.

It asserts correctly.

Thank you.

```java
@Test
    public void testContainsValue() throws ConfigurationException {
        final DatabaseConfiguration config = helper.setUpConfig();
        config.addPropertyDirect("test", "test1");

        assertTrue(config.containsValue("test1"));
    }
```

**Comment by garydgregory** [comment]
> @garydgregory , thank you for the feedback and providing the documentation.
> 
> I must rethink my profile description.

It's hard to read people online but it appears you have good sense of humor 😀

**Comment by rikkarth** [comment]
> > @garydgregory , thank you for the feedback and providing the documentation.
> > I must rethink my profile description.
> 
> It's hard to read people online but it appears you have good sense of humor 😀

A bit of sense of humour and humility is always nice to have.

I'm very grateful that you and @Claudenw have taken the time to look at my PR, thank you.

As for the compatibility issue, I don't think I'm breaking compatibility by adding a new method to the API, so that means I don't have to bump the version from 2.x to 3.x..

I would like to ask, how are you bumping the version since from past PRs I've read I don't see any change to the pom.

**Comment by rikkarth** [comment]
Ok so I believe this is the reason why I broke compatibility.

```
However, adding overloaded methods has the potential to change method resolution and thus change the signatures of the method call sites in the resulting class file. Whether or not such a change is problematic with respect to source compatibility depends on what semantics are required and how the different overloaded methods operate on the same inputs, which interacts with behavioral equivalence notions. For example, consider adding to Lib an overloading of foo which takes an int:

// Behaviorally equivalent source compatible change
public final class Lib {
  public double foo(double d) {
    return d * 2.0;
  }

  // New overloading
  public double foo(int i) {
    return i * 2.0;
  }
}
```

**Comment by garydgregory** [comment]
@rikkarth
You need to understand how to not break binary compatibility: Please see "13.5.6. Interface Method Declarations" in the JLS at the link I've already provided.

**Comment by rikkarth** [comment]
> @rikkarth You need to understand how to not break binary compatibility: Please see "13.5.6. Interface Method Declarations" in the JLS at the link I've already provided.

Thank you so much for pointing me in the right direction. First time contributing to a project with this specification, but after reading all this it makes a lot of sense.

I would like to ask if you don't mind looking at my default implementation.

I think if this method is not implemented by the Class implementing ImmutableConfiguration, the caller should get an Exception with this message.

Or should it just return false?

```java
    /**
     * Returns true if this configuration contains one or more keys to this value. This operation is more expensive than
     * the {@link #containsKey containsKey} method.
     *
     * @param value value whose presence in this configuration is to be tested
     * @return {@code true} if this configuration maps one or more keys to the specified value
     * @throws NullPointerException if the value is {@code null}
     * @throws UnsupportedOperationException if this method is not supported by the implementing class.
     */
    default boolean containsValue(String value) {
        throw new UnsupportedOperationException(String.format("Class %s does not support this method.", this.getClass()));
    }
```

**Comment by garydgregory** [comment]
@rikkarth 
If you think your PR is ready for review, then set it to "Ready for review" and I'll take a look over the next day or two.

**Comment by rikkarth** [comment]
Thank you for taking the time to look at my PR @garydgregory and @Claudenw .

I've submitted my PR for review as instructed.

Have a great day!

File: src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java
Comment (garydgregory) [inline src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java:427]: - Replace API calls of this pattern with simply:
`Objects.requireNonNull(value, "value");`
See the rest of the code base.
- Why does this check need to happen here at all? After all I don't get an NPE when I ask for a null value from a `HashMap`, for example ` new HashMap().containsValue(null)` returns false.

File: src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java
Comment (garydgregory) [inline src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java:459]: New elements that are public or private should be documented with a Javadoc `@since` tag.

File: src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java
Comment (garydgregory) [inline src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java:1558]: Use `final` where you can, here, `keys` can be `final`.

File: src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java
Comment (garydgregory) [inline src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java:1559]: See my previous comment RE `Objects.requireNonNull()`.
Same comment RE null checking in the first place. Why not return false?

File: src/main/java/org/apache/commons/configuration2/DatabaseConfiguration.java
Comment (garydgregory) [inline src/main/java/org/apache/commons/configuration2/DatabaseConfiguration.java:491]: No need for extra whitespace (blank lines). If you want to call out a code fragment, then you can use an inline `//` comment to alert the reader to something specific.

File: src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java
Comment (garydgregory) [inline src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java:66]: Be consistent in your Javadoc comments. In this case, "contains" methods that return a boolean are "testing" for the presence of something, so you can start the comment with "Tests" like "Tests whether this or that happens".

File: src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java
Comment (garydgregory) [inline src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java:73]: Missing Javadoc since tag.

File: src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java
Comment (garydgregory) [inline src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java:74]: Can this method be implemented in terms of other interface methods instead of throwing an exception?

File: src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java
Comment (garydgregory) [inline src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java:76]: The message is redundant and bad, it read like "Class class com.example.Foo does not...".
You must have not tested this case. A simpler message could just say `getClass().getName() + ".myMethod"`

File: src/test/java/org/apache/commons/configuration2/NonCloneableConfiguration.java
Comment (garydgregory) [inline src/test/java/org/apache/commons/configuration2/NonCloneableConfiguration.java:48]: Javadoc

File: src/test/java/org/apache/commons/configuration2/TestAbstractConfiguration.java
Comment (garydgregory) [inline src/test/java/org/apache/commons/configuration2/TestAbstractConfiguration.java:83]: Use `final`.

File: src/test/java/org/apache/commons/configuration2/TestAbstractConfiguration.java
Comment (garydgregory) [inline src/test/java/org/apache/commons/configuration2/TestAbstractConfiguration.java:91]: No need to blank lines. See my previous comment.

File: src/test/java/org/apache/commons/configuration2/TestAbstractConfiguration.java
Comment (garydgregory) [inline src/test/java/org/apache/commons/configuration2/TestAbstractConfiguration.java:82]: The style of the component is to use `public` test methods prefixed with "test" in the name and camel-case. Look at other tests for examples.

**Review by garydgregory — CHANGES_REQUESTED** [review]
@rikkarth 
Thank you for your PR.
Please see my comments.

File: src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java
Comment (rikkarth) [reply src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java:76]: Hey Greg, you are right

I did in fact not test this method since I couldn't think of a way to test this in a class that doesn't implement the method without creating a dummy class for it.

Do you think I should just adapt the message like you suggested?

Thank you.

Comment (garydgregory) [reply src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java:76]: Yes.

File: src/test/java/org/apache/commons/configuration2/TestAbstractConfiguration.java
Comment (rikkarth) [reply src/test/java/org/apache/commons/configuration2/TestAbstractConfiguration.java:82]: Will correct.

File: src/test/java/org/apache/commons/configuration2/TestAbstractConfiguration.java
Comment (rikkarth) [reply src/test/java/org/apache/commons/configuration2/TestAbstractConfiguration.java:91]: Corrected.

File: src/test/java/org/apache/commons/configuration2/TestAbstractConfiguration.java
Comment (rikkarth) [reply src/test/java/org/apache/commons/configuration2/TestAbstractConfiguration.java:83]: Done.

File: src/main/java/org/apache/commons/configuration2/DatabaseConfiguration.java
Comment (rikkarth) [reply src/main/java/org/apache/commons/configuration2/DatabaseConfiguration.java:491]: Corrected.

File: src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java
Comment (rikkarth) [reply src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java:1558]: Will do.

File: src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java
Comment (rikkarth) [reply src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java:427]: containsValue delegates to `containsKeyInternal()`.

`containsKeyInternal()` implementation can be different depending on the implementing class.

If null is unavoidable we should always encourage safe null handling practices, such as using Objects.requireNonNull or conditional checks, to minimize the risk of (unexpected, unmanaged or uncontrolled) null pointer exceptions.

While `new HashMap().containsValue(null)` returns false, this implementation is inspired on `java.util.HashTable.contains()` and not HashMap which is also `java.util.Properties` implementation (extends HashTable).

```java
    // HashTable contains Implementation checks null at entry point
    public synchronized boolean contains(Object value) {
        if (value == null) {
            throw new NullPointerException();
        }

        Entry<?,?> tab[] = table;
        for (int i = tab.length ; i-- > 0 ;) {
            for (Entry<?,?> e = tab[i] ; e != null ; e = e.next) {
                if (e.value.equals(value)) {
                    return true;
                }
            }
        }
        return false;
    }
```

I find this to be a much more coherent implementation since it doesn't fail silently and explicitely warns the user that the input value is null.

The reason for the "custom" message is so we can control exactly how containsValue fails.

What do you think?

File: src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java
Comment (rikkarth) [reply src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java:459]: How about protected ones? (sorry I know you didn't mention but just want to make sure)

File: src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java
Comment (rikkarth) [reply src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java:66]: Thank you for the feedback. I have adjusted this.

File: src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java
Comment (rikkarth) [reply src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java:73]: Done.

File: src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java
Comment (rikkarth) [reply src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java:74]: Ok, so I gave an actual default implementation to `containsValue`  in `ImmutableConfiguration` and tested it with that DummyClass that I didn't want to create to test it and see if it worked.

And it worked.

image

image

image

File: src/test/java/org/apache/commons/configuration2/NonCloneableConfiguration.java
Comment (rikkarth) [reply src/test/java/org/apache/commons/configuration2/NonCloneableConfiguration.java:51]: Done.

File: src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java
Comment (Claudenw) [inline src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java:427]: I think the `Objects.requiredNull` should be removed since this is an AbstractConfiguration and my actually be implementing a configuration that allows nulls.  It may make sense in final classes to check for null but not here.

File: src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java
Comment (Claudenw) [reply src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java:459]: My approach is: if you are writing a javadoc add the `@since` annotation.

File: src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java
Comment (Claudenw) [reply src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java:1559]: See my previous comment re `Objects.requireNonNull()`

**Review by Claudenw — CHANGES_REQUESTED** [review]
Thanks for your work on this.  It is looking better.

File: src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java
Comment (rikkarth) [reply src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java:459]: Thank you I wasn't aware of this rule until this PR, thank you again for clarifying will do that from now on.

File: src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java
Comment (rikkarth) [reply src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java:427]: I agree, that makes sense, since `containsValue` delegates to containsValueInternal there could be an implementation that allows null values.

Will remove.

Comment (rikkarth) [reply src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java:427]: If every configuration can have a different implementation for containsValue it means that every configuration must be tested for their (existing) implementation of containsValue.

Is it ok if I do it all in this PR?

Comment (rikkarth) [reply src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java]: After discussion with @Claudenw I have removed the null validation. Please see comments.

Thank you :)

**Review by garydgregory — CHANGES_REQUESTED** [review]
Hello @rikkarth 
- Please rebase on git master to pick up the latest Checkstyle configuration and fixes.
- See my comments in this PR
- Run `mvn` by itself to run all build checks

TY!

File: src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java
Comment (rikkarth) [reply src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java:424]: Do you mean the version value? Or the tag?

File: src/test/java/org/apache/commons/configuration2/TestAbstractConfiguration.java
Comment (rikkarth) [reply src/test/java/org/apache/commons/configuration2/TestAbstractConfiguration.java:84]: Will be removed.

File: src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java
Comment (garydgregory) [reply src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java]: New public and protected method should have a Javadoc since tag. Semantic versioning makes the next version 2.11.0.

**Comment by rikkarth** [comment]
> Hello @rikkarth
> 
> * Please rebase on git master to pick up the latest Checkstyle configuration and fixes.
> * See my comments in this PR
> * Run `mvn` by itself to run all build checks
> 
> TY!

Done as requested.

While changing all the since tags and adjusting tests signature names I was wondering how dull must have been for you to review this, sorry about that. In future PRs I will know better the design, patterns and style applied to this component from the experience acquired in this first PR.

Thank you for your time again @garydgregory

File: .github/workflows/coverage.yml
Comment (garydgregory) [inline .github/workflows/coverage.yml:45]: The PR should not contain this change.

File: .github/workflows/maven.yml
Comment (garydgregory) [inline .github/workflows/maven.yml:30]: The PR should not contain this change.

**Review by garydgregory — CHANGES_REQUESTED** [review]
The PR should not contain a bunch of these changed files. Please rollback those changes to only contain the required changes. ty!

**Comment by rikkarth** [comment]
Done as instructed.

**Review by garydgregory — CHANGES_REQUESTED** [review]
@rikkarth 
Something is still messed up in this PR. It touches 67 files! It looks like it contains unrelated changes from master. Something must have gone wrong in whatever git commands you used. The PR should only contain _your_ changes.
TY!

**Comment by rikkarth** [comment]
I'll fix it. Had to change computers and my gpg was not signing commits.

**Comment by rikkarth** [comment]
I believe it's done now @garydgregory :) had to do some git gymnastics «

File: .github/workflows/coverage.yml
Comment (garydgregory) [inline .github/workflows/coverage.yml:45]: This change does not belong in this PR.

File: .github/workflows/maven.yml
Comment (garydgregory) [inline .github/workflows/maven.yml:30]: This change does not belong in this PR.

**Review by garydgregory — CHANGES_REQUESTED** [review]
Two changes do not belong in this PR. See my comments. TY!

**Comment by rikkarth** [comment]
I was under the wrong impression you wanted those changes, but I reverted them now!

**Pull request merged**
